include/cppkafka/utils/buffered_producer.h

```diff
@@ -661,25 +672,44 @@ bool BufferedProducer<BufferType, Allocator>::flush(std::chrono::milliseconds ti
     if (preserve_order) {
         CounterGuard<size_t> counter_guard(flushes_in_progress_);
         QueueType flush_queue; // flush from temporary queue
+        QueueType retry_flush_queue; // flush from temporary retry queue
         {
             std::lock_guard<std::mutex> lock(mutex_);
+            std::swap(retry_messages_, retry_flush_queue);
             std::swap(messages_, flush_queue);
         }
+        auto queue_flusher = [this](QueueType& queue)->bool
+        {
+            if (!queue.empty()) {
+                sync_produce(queue.front());
+                queue.pop_front();
+                return true;
+            }
+            return false;
+        };
         auto remaining = timeout;
         auto start_time = std::chrono::high_resolution_clock::now();
         do {
-            sync_produce(flush_queue.front());
-            flush_queue.pop_front();
+            if (!queue_flusher(retry_flush_queue) && !queue_flusher(flush_queue)) {
+                break;
+            }
             // calculate remaining time
             remaining = timeout - std::chrono::duration_cast<std::chrono::milliseconds>
                 (std::chrono::high_resolution_clock::now() - start_time);
-        } while (!flush_queue.empty() && (remaining.count() > 0));
+        } while (remaining.count() > 0);
         
         // Re-enqueue remaining messages in original order
-        if (!flush_queue.empty()) {
-            std::lock_guard<std::mutex> lock(mutex_);
-            messages_.insert(messages_.begin(), std::make_move_iterator(flush_queue.begin()), std::make_move_iterator(flush_queue.end()));
-        }
+        auto re_enqueuer = [this](QueueType& src_queue, QueueType& dst_queue)->void
+        {
+            if (!src_queue.empty()) {
+                std::lock_guard<std::mutex> lock(mutex_);
```

Comment:
Since the logic for flushing/re-enqueuing has now been made generic into lambdas, I wonder if it's not better to have 2 mutexes (one for each queue) and just pass it in the lambda as well as a parameter. There is no point in blocking operations on one queue when the other one is being accessed. Especially since the `retry_queue_` is populated primarily from the rdkafka callback, you don't want to block it unnecessarily when the main queue is being used.